egs/wsj/s5/local/chain/tuning/run_tdnn_1h.sh

```diff
@@ -0,0 +1,362 @@
+  echo "$0: creating lang directory $lang with chain-type topology"
+  # Create a version of the lang/ directory that has one state per phone in the
+  # topo file. [note, it really has two states.. the first one is only repeated
+  # once, the second one has zero or more repeats.]
+  if [ -d $lang ]; then
+    if [ $lang/L.fst -nt data/lang/L.fst ]; then
+      echo "$0: $lang already exists, not overwriting it; continuing"
+    else
+      echo "$0: $lang already exists and seems to be older than data/lang..."
+      echo " ... not sure what to do.  Exiting."
+      exit 1;
+    fi
+  else
+    cp -r data/lang $lang
+    silphonelist=$(cat $lang/phones/silence.csl) || exit 1;
+    nonsilphonelist=$(cat $lang/phones/nonsilence.csl) || exit 1;
+    # Use our special topology... note that later on may have to tune this
+    # topology.
+    steps/nnet3/chain/gen_topo.py $nonsilphonelist $silphonelist >$lang/topo
+  fi
+fi
+
+if [ $stage -le 13 ]; then
+  # Get the alignments as lattices (gives the chain training more freedom).
+  # use the same num-jobs as the alignments
+  steps/align_fmllr_lats.sh --nj 100 --cmd "$train_cmd" ${lores_train_data_dir} \
+    data/lang $gmm_dir $lat_dir
+  rm $lat_dir/fsts.*.gz # save space
+fi
+
+if [ $stage -le 14 ]; then
+  # Build a tree using our new topology.  We know we have alignments for the
+  # speed-perturbed data (local/nnet3/run_ivector_common.sh made them), so use
+  # those.  The num-leaves is always somewhat less than the num-leaves from
+  # the GMM baseline.
+   if [ -f $tree_dir/final.mdl ]; then
+     echo "$0: $tree_dir/final.mdl already exists, refusing to overwrite it."
+     exit 1;
+  fi
+  steps/nnet3/chain/build_tree.sh \
+    --frame-subsampling-factor 3 \
+    --context-opts "--context-width=2 --central-position=1" \
+    --cmd "$train_cmd" 3500 ${lores_train_data_dir} \
+    $lang $ali_dir $tree_dir
+fi
+
+
+if [ $stage -le 15 ]; then
+  mkdir -p $dir
+  echo "$0: creating neural net configs using the xconfig parser";
+
+  num_targets=$(tree-info $tree_dir/tree |grep num-pdfs|awk '{print $2}')
+  learning_rate_factor=$(echo "print(0.5/$xent_regularize)" | python)
+  tdnn_opts="l2-regularize=0.01 dropout-proportion=0.0 dropout-per-dim-continuous=true"
+  tdnnf_opts="l2-regularize=0.01 dropout-proportion=0.0 bypass-scale=0.66"
+  linear_opts="l2-regularize=0.01 orthonormal-constraint=-1.0"
+  prefinal_opts="l2-regularize=0.01"
+  output_opts="l2-regularize=0.005"
+
+  mkdir -p $dir/configs
+  cat <<EOF > $dir/configs/network.xconfig
+  input dim=100 name=ivector
+  input dim=40 name=input
+
+  idct-layer name=idct input=input dim=40 cepstral-lifter=22 affine-transform-file=$dir/configs/idct.mat
+
+  no-op-component name=delta_1 input=Scale(-1.0, Offset(idct, -1))
+  no-op-component name=delta_2 input=Scale(1.0, Offset(idct, 1))
+  no-op-component name=delta_delta_1 input=Scale(-2.0, Offset(idct, 0))
+  no-op-component name=delta_delta_2 input=Scale(1.0, Offset(idct, -2))
```

Comment:
Hm, I think before I merge this I'd like to have some work done on the scripts, to make this much more compact.
Firstly, please don't use any more no-op-components than are absolutely necessary, they lead to redundant matrix copies.  I suspect no more than 2 of those, at most, are needed.  In fact, the following would compile with the least hassle:
 `dim-range-node name=idct_copy1 input=idct offset=0 dim=40`
 `dim-range-node name=idct_copy2 input=idct offset=0 dim=40`
(the syntax may be off slightly):
and then whenever you get an error due to something appearing in the sum with 2 different scales, just use a different copy.  And I'd like this to all be put in an xconfig layer so that we don't need to bulk up the config files.
If you're not sure how, talk to one of the others.  You could copy-and-modify something simple like the `no-op-component` in trivial_layers.py; you can leave it in trivial layers.py; you'll need to add a line to parser.py.

Reply:
Okay, I'll make those changes.

Reply:
Should I just fix the scale as 1.0 or have it as a parameter for the layer (along with a parameter to choose if we want to use batchnorm)? That would be more general, but I suppose in that case I would need to have several NoOpComponents.

Reply:
In training, delta feature is chunk-based, while in testing, it is utt-based.

Reply:
@ishine: that's not really true as any required left/right context is stored with the chunks.